src/cmdliner.ml

```diff
@@ -245,32 +245,139 @@ module Term = struct
+      | Term (al, f) -> (Term (al, f), term_add_args (al, f) info)
+      | Group subs -> (Group (List.map add_args subs), info)
+
+    let (>>=) res f =
+      match res with
+      | Error e -> Error e
+      | Ok x -> f x
+
+    let parse_arg_cmd = function
+    | [] -> Error `No_args
+    | cmd :: args ->
+        if String.length cmd >= 1 && cmd.[0] = '-' then
+          Error `No_args
+        else
+        Ok (cmd, args)
+
+    let cmd_name (_, info) = Cmdliner_info.term_name info
+
+    let one_of (cmd, (choices : _ t list), path, args) =
+      let index =
+        let add acc c =
+          let name = cmd_name c in
+          match Cmdliner_trie.add acc name c with
+          | `New t -> t
+          | `Replaced (c', _) ->
+              let flip (x, y) = (y, x) in
+              invalid_arg (err_multi_cmd_def name (flip c) (flip c'))
+        in
+        List.fold_left add Cmdliner_trie.empty choices
+      in
+      match Cmdliner_trie.find index cmd with
+      | `Ok (choice, info) -> Ok ((choice, info), choices, info :: path, args)
+      | `Not_found ->
+          let all = Cmdliner_trie.ambiguities index "" in
+          let hints = Cmdliner_suggest.value cmd all in
+          Error (`Invalid_command (cmd, path, choices, hints))
+      | `Ambiguous ->
+          let ambs = Cmdliner_trie.ambiguities index cmd in
+          let ambs = List.sort compare ambs in
+          Error (`Ambiguous (cmd, path, ambs))
+
+    let try_one_of choices path args =
+      match parse_arg_cmd args with
+      | Ok (cmd, args) -> one_of (cmd, choices, path, args)
+      | Error `No_args -> Error (`No_args (path, choices))
+
+    let rec try_choose_term choices path args =
+      try_one_of choices path args >>= choose_term
+
+    and choose_term ((t, info), choices, path, args) =
+      match t with
+      | Term t -> Ok ((t, info), choices, path, args)
+      | Group subs -> try_choose_term subs path args
+
+    let choose_term main choices args =
+      let path = [snd main] in
+      match parse_arg_cmd args with
+      | Error `No_args -> Ok (main, choices, path, args)
+      | Ok (cmd, args) -> one_of (cmd, choices, path, args) >>= choose_term
+
+    let eval
+        ?help:(help_ppf = Format.std_formatter)
+        ?err:(err_ppf = Format.err_formatter)
+        ?(catch = true) ?(env = env_default) ?(argv = Sys.argv) main choices =
+    let choices_f = List.map add_args choices in
     let to_term_f ((al, f), ti) = Cmdliner_info.term_add_args ti al, f in
-    let choices_f = List.rev_map to_term_f choices in
+    let main_args = fst main in
     let main_f = to_term_f main in
-    let choices = List.rev_map fst choices_f in
     let main = fst main_f in
-    match choose_term main_f choices_f (remove_exec argv) with
-    | Error err ->
-        let ei = Cmdliner_info.eval ~term:main ~main ~choices ~env in
+    match choose_term (main_args, (fst main_f)) choices_f (remove_exec argv) with
+    | Error (`No_args (path, choices)) ->
```

Comment:
Could we make this configurable, so that the user can supply a default behavior if no subcommand is given? We'd need this for  the behavior planned here: https://github.com/ocaml/dune/issues/4367#issuecomment-815056360

Reply:
IIRC. I'd rather not. See the discussion starting [here](https://github.com/dbuenzli/cmdliner/issues/24#issuecomment-495198239).

Reply:
It seems common to have a natural default behavior for sub commands (see `git branch` or `git remote` or `opam switch` etc.) (as you pointed out on linked conversation, in which, it seems, you were undecided at the time). It's also worth noting that @jeremiedimino, who had suggested the limitation, has actually requested precisely this behavior in the linked comment!

I wonder if the seeming recurrence of this pattern might cause a reconsideration on this point? :)

Reply:
I'm actually reconsidering. 

I think in the general case selecting a nested command named `cmd : string list` should work as follows. Given command line arguments `args : string list` without the exec name:

1. Let `sel : string list` be the list of all arguments of `args` not prefixed by a `-` and occuring before a potential `--`.
2. If the `cmd` is a prefix of `sel` (and there's no longer command matching `sel`) then `path` is the nested command to use. `cmd` is then removed from `args` and the result is parsed using `path`'s term.

This allows to have a default command at any level (if that command wants positional arguments they have to be specified after a `--`). Besides in contrast to the current behaviour for multi commands which requires the first argument of `args` to be the name of the sub command. It allows to specify options before it, as long as those do not used the glued form (like `-o file` or `--output file`). 

The end user can still end up being confused in many ways while refining cli invocations, but that's a bit in the dna of this poor interface medium and the behaviour seems not too hard to explain and understand.

Reply:
Of course that doesn't work :-) (e.g. `js_of_ocaml -o bla.js` breaks, it tries to parse `bla.js` as a command). 

Even though I find that annoying, I think we are forced to require that no optional arguments occur before subcommand specifications (as was always the case in `cmdliner`).